### DISCUSSION TITLE
Proposal: Params function source gnerator

### DISCUSSION THREAD

**Discussion by JPaja** [dotnet/csharplang#4619] [General]
When i watched Dotnetos Conference talk about optimizations by Stephen Toub,
I noticed that he mentioned how adding new instances of method `Task.WhenAny` with const parameter lengths
improved speed by skipping array allocation on method call.

https://youtu.be/Ehvr9wXJNEM?t=1151
https://github.com/dotnet/runtime/pull/34288

So my idea would be to create source generator that would automatically construct and optimize those methods,
and  give warning and leave it as is if its not possible to do so.

for example
```csharp
[ConstructParams(paramDepth: 2)]
void PrintArgs(params string[] args)
{
  foreach(var arg in args)
     Console.WriteLine(arg);
}
```
would generate 2 methods behind the scenes
```csharp
void PrintArgs(string arg1)
{
     Console.WriteLine(arg1);
}

void PrintArgs(string arg1, string arg2)
{
     Console.WriteLine(arg1);
     Console.WriteLine(arg2);
}
```

**Comment by BreyerW** [comment]
There is already an idea to allow readonlyspan with params. This should give almost same performance since spans do not allocate (you only have to pay for span creation which is much cheaper). However team already has a lot on their plate so this idea ended up in backlog....

[https://github.com/dotnet/csharplang/blob/main/proposals/format.md](https://github.com/dotnet/csharplang/blob/main/proposals/format.md)

**Reply by TonyValenti** [reply]
Related to this, I would love a source generator that would create params methods for all methods that end with an IEnumerable type.

**Comment by canton7** [comment]
I don't think there's any automatic way to generate the body of these methods though. Your example just repeats Console.WriteLine, but real-world examples are rarely this simple. String.Format for example had dedicated infrastructure to accept up to 3 explicit params, and the Task.WhenAny overload which takes 2 tasks has a dedicated 2-task-specific implementation.
